csrc/scheduler/multi_matmul.cpp

```diff
@@ -1006,16 +1006,146 @@ class MultipleMatmulScheduler {
+    // prologue (generally) in registers. We typically refer to the register
+    // buffer that is loaded from operand A's smem buffer using ldmatrix as
+    // "acr". This is the beginning of the register prologue region for that
+    // operand. The end of that region is the first input to the MmaOp
+    // expression, which we typically refer to as "ab". There is some special
+    // handling of acr but otherwise we schedule ab and propagate backward
+    // along this prologue region.
+    auto schedulePrologueBranch = [&](const std::vector<TensorView*>&
+                                          smem_stores,
+                                      const std::vector<TensorView*>&
+                                          smem_loads,
+                                      std::vector<TensorView*>& mma_inputs,
+                                      MmaOperand operand_type) {
+      NVF_ERROR(smem_stores.size() == smem_loads.size());
+      // TODO: we should not assume that each operand is used in only a single
+      // mma op
+      NVF_ERROR(mma_results_.size() >= smem_loads.size());
+      // We will save abs_ and bbs_ here for later use
+      // TODO: save all register prologue tensors instead to a new vector called
+      // prologue_register_tensors_
+      NVF_ERROR(mma_inputs.empty());
+      for (TensorView* mma_result : mma_results_) {
+        MmaOp* mma = dynamic_cast<MmaOp*>(mma_result->definition());
+        NVF_ERROR(mma != nullptr);
+        TensorView* mma_input = nullptr;
+        if (operand_type == MmaOperand::A) {
+          mma_input = mma->inA()->as<TensorView>();
+        } else if (operand_type == MmaOperand::B) {
+          mma_input = mma->inB()->as<TensorView>();
+        }
+        NVF_ERROR(mma_input != nullptr);
+        mma_inputs.push_back(mma_input);
+      }
+
+      scheduleMmaOperands(mma_inputs, operand_type);
+
+      // Propagate backward from all mma_results to smem_stores
+
+      for (TensorView* mma_input : mma_inputs) {
+        // Schedule mma_input, since we know it has the broadcast dimension M or
+        // N, whereas the smem read might not
+        moveInnerBroadcastLeft(mma_input);
+        mma_input->applyMmaSwizzle(operand_type);
+        scheduler_utils::BoundedDirectionalTransformPropagator::backward(
+            mma_input,
+            -1,
+            smem_stores,
+            scheduler_utils::BoundedDirectionalTransformPropagator::Options()
+                .propagateParallelType());
+      }
+      // Find smem loads that are mma inputs and save them
+      std::unordered_set<TensorView*> smem_load_mma_inputs;
+      for (TensorView* smem_load : smem_loads) {
+        // Insert only if smem_load is also in mma_inputs
+        bool is_mma_input =
+            std::find(mma_inputs.begin(), mma_inputs.end(), smem_load) !=
+            mma_inputs.end();
+        if (is_mma_input) {
+          smem_load_mma_inputs.insert(smem_load);
+        }
+        if (!is_mma_input) {
```

Comment:
This logic looks odd to me. I believe it is reqiured because in `applyMmaSwizzle` and `scheduleOperandRead`, there are logics that does things differently for input vs non-input. Could we refactor `applyMmaSwizzle` to make all of `applyMmaSwizzle`, `scheduleOperandRead` and this function simpler? Let's do this in a separate PR. This  PR as-is looks good.

Reply:
I agree. This felt ugly to write, but it corresponds to the condition here https://github.com/NVIDIA/Fuser/blob/027cd0f5b5e723ab5c57c75da9ec8852e0e23285/csrc/scheduler/matmul.cpp#L1114
I'll look at cleaning this up in a separate PR, as you suggest.